### PR TITLE
core: introduce Inherited.upcast to fix warning in javadsl.Http

### PR DESCRIPTION
**Pull request by jrudolph**
An alternative to #2894 for this particular case.

### CODE REVIEW

File: akka-http-core/src/main/scala/akka/http/impl/util/JavaMapping.scala [hunk 1]
```diff
@@ -180,6 +180,9 @@ private[http] object JavaMapping {
   class Inherited[J <: AnyRef, S <: J](implicit classTag: ClassTag[S]) extends JavaMapping[J, S] {
     def toJava(scalaObject: S): J = scalaObject
     def toScala(javaObject: J): S = cast[S](javaObject)
+
+    import scala.language.higherKinds
+    def upcast[F[+_]](f: F[J]): F[S] = f.asInstanceOf[F[S]]
```
Comment (jrudolph) [inline akka-http-core/src/main/scala/akka/http/impl/util/JavaMapping.scala]: or should it be `downcast`?

Comment (raboof) [reply akka-http-core/src/main/scala/akka/http/impl/util/JavaMapping.scala]: I guess it's a (safe) downcast https://en.wikipedia.org/wiki/Downcasting

File: akka-http-core/src/main/scala/akka/http/javadsl/Http.scala [hunk 2]
```diff
@@ -40,7 +40,7 @@ class Http(system: ExtendedActorSystem) extends akka.actor.Extension {
 
   import language.implicitConversions
   private implicit def completionStageCovariant[T, U >: T](in: CompletionStage[T]): CompletionStage[U] = in.asInstanceOf[CompletionStage[U]]
-  private implicit def javaModelIsScalaModel[J <: AnyRef, S <: J](in: Future[J])(implicit ev: JavaMapping.Inherited[J, S]): Future[S] = in.asInstanceOf[Future[S]]
+  private implicit def javaModelIsScalaModel[J <: AnyRef, S <: J](in: Future[J])(implicit ev: JavaMapping.Inherited[J, S]): Future[S] = ev.upcast(in)
```
Comment (raboof) [inline akka-http-core/src/main/scala/akka/http/javadsl/Http.scala]: nice, more explicit

File: akka-http-core/src/main/scala/akka/http/impl/util/JavaMapping.scala (hunk 1 above)
Comment (jrudolph) [reply akka-http-core/src/main/scala/akka/http/impl/util/JavaMapping.scala]: Woops, thanks, fixed.

**Pull request merged**